A bug about MatchFunc with indexed field

Hi,

This awesome project is very helpful for me. Thanks a lot.

When I added index tag to Name, got a bug:

=== RUN   TestMatchFunc
--- FAIL: TestMatchFunc (0.03s)
	bolthold_test.go:27: gob: decoding into local type *bolthold.MatchFunc, received remote type string

test code:
```
func TestMatchFunc(t *testing.T) {
	store, err := bolthold.Open("test.db", 0666, nil)
	if err != nil {
		t.Error(err)
	}
	defer store.Close()
	if err := store.Insert("key", &Item{
		Name:    "Test Name",
		Created: time.Now(),
	}); err != nil {
		t.Error(err)
	}
	result := []Item{}
	if err := store.Find(&result, bolthold.Where("Name").MatchFunc(func(ra *bolthold.RecordAccess) (bool, error) {
		t.Logf("%T,%T", ra.Field(), ra.Record())
		return ra.Field() == "Test Name", nil
	})); err != nil {
		t.Error(err)
	} else {
		t.Log(result)
	}
}

type Item struct {
	Name    string    `boltholdIndex:"Name"`
	Created time.Time `boltholdIndex:"Created"`
}
```

--- a/find_test.go
+++ b/find_test.go
@@ -418,6 +418,19 @@ var tests = []test{
 		query:  bolthold.Where(bolthold.Key).Gt(testData[10].Key).Limit(5),
 		result: []int{11, 12, 13, 14, 15},
 	},
+	test{
+		name: "Issue #8 - Function Field on index",
+		query: bolthold.Where("Category").MatchFunc(func(ra *bolthold.RecordAccess) (bool, error) {
+			field := ra.Field()
+			_, ok := field.(string)
+			if !ok {
+				return false, fmt.Errorf("Field not a string, it's a %T!", field)
+			}
+
+			return !strings.HasPrefix(field.(string), "veh"), nil
+		}),
+		result: []int{2, 4, 5, 7, 8, 9, 10, 12, 13, 14, 15, 16},
+	},
 }
 
 func insertTestData(t *testing.T, store *bolthold.Store) {
@@ -745,5 +758,26 @@ func TestSlicePointerResult(t *testing.T) {
 		if len(result) != count {
 			t.Fatalf("Expected %d, got %d", count, len(result))
 		}
+	})
+}
+
+func TestKeyMatchFunc(t *testing.T) {
+	testWrap(t, func(store *bolthold.Store, t *testing.T) {
+		defer func() {
+			if r := recover(); r == nil {
+				t.Fatalf("Running matchFunc against Key query did not panic!")
+			}
+		}()
+
+		var result []ItemTest
+		_ = store.Find(&result, bolthold.Where(bolthold.Key).MatchFunc(func(ra *bolthold.RecordAccess) (bool, error) {
+			field := ra.Field()
+			_, ok := field.(string)
+			if !ok {
+				return false, fmt.Errorf("Field not a string, it's a %T!", field)
+			}
+
+			return strings.HasPrefix(field.(string), "oat"), nil
+		}))
 	})
 }

--- a/query.go
+++ b/query.go
@@ -40,6 +40,7 @@ type Query struct {
 
 	badIndex   bool
 	currentRow interface{}
+	dataType   reflect.Type
 	tx         *bolt.Tx
 
 	limit int
@@ -284,6 +285,9 @@ func (r *RecordAccess) SubQuery(result interface{}, query *Query) error {
 
 // MatchFunc will test if a field matches the passed in function
 func (c *Criterion) MatchFunc(match MatchFunc) *Query {
+	if c.query.currentField == Key {
+		panic("Match func cannot be used against Keys, as the Key type is unknown at runtime, and there is no value compare against.")
+	}
 	return c.op(fn, match)
 }
 
@@ -292,11 +296,34 @@ func (c *Criterion) test(testValue interface{}, encoded bool) (bool, error) {
 	var value interface{}
 	if encoded {
 		if len(testValue.([]byte)) != 0 {
-			// used with keys
-			value = reflect.New(reflect.TypeOf(c.value)).Interface()
-			err := decode(testValue.([]byte), value)
-			if err != nil {
-				return false, err
+			if c.operator == fn {
+				// with matchFunc, their is no value type to assume a type from, so we need to get it
+				// from the current field being tested
+				// This is not possible with Keys, so defining a matchFunc query on a Key panics
+
+				fieldType, ok := c.query.dataType.FieldByName(c.query.index)
+				if !ok {
+					return false, fmt.Errorf("Current row does not contain the field %s which has been indexed.",
+						c.query.index)
+				}
+				value = reflect.New(fieldType.Type).Interface()
+				err := decode(testValue.([]byte), value)
+				if err != nil {
+					return false, err
+				}
+
+				// in order to decode, we needed a pointer, but matchFunc is expecting the original
+				// type, so we need to get an interface of the actual element, not the pointer
+				value = reflect.ValueOf(value).Elem().Interface()
+
+			} else {
+				// used with keys
+				value = reflect.New(reflect.TypeOf(c.value)).Interface()
+				err := decode(testValue.([]byte), value)
+				if err != nil {
+					return false, err
+				}
+
 			}
 		}
 
@@ -446,6 +473,8 @@ func runQuery(tx *bolt.Tx, dataType interface{}, query *Query, retrievedKeys key
 	for reflect.TypeOf(tp).Kind() == reflect.Ptr {
 		tp = reflect.ValueOf(tp).Elem().Interface()
 	}
+
+	query.dataType = reflect.TypeOf(tp)
 
 	iter := newIterator(tx, storer.Type(), query)
 

